src/lib/FreeC/LiftedIR/Converter/Type.hs

```diff
@@ -0,0 +1,105 @@
+-- | Implements the IR to lifted IR translation, which applies the monadic lifting
+--   as described by Abel et al. in "Verifying Haskell Programs Using Constructive
+--   Type Theory".
+
+module FreeC.LiftedIR.Converter.Type
+  ( liftFuncArgTypes
+  , liftConArgType
+    -- * Translations
+  , liftType
+  , liftType'
+  )
+where
+
+import qualified FreeC.IR.Syntax               as IR
+import           FreeC.IR.SrcSpan               ( SrcSpan(NoSrcSpan) )
+import qualified FreeC.LiftedIR.Syntax         as LIR
+import           FreeC.Util.SnocList            ( SnocList(Nil) )
+
+-- | Converts the argument types of a function.
+liftFuncArgTypes
+  :: Maybe Int -- ^ Index of the decreasing argument for recursive functions.
+  -> [IR.Type] -- ^ The argument types.
+  -> [LIR.Type]
+liftFuncArgTypes = maybe liftNonRecFuncArgTypes liftRecFuncArgTypes
+
+-- | Converts the argument types of a non recursive function using @convertType@.
+liftNonRecFuncArgTypes :: [IR.Type] -> [LIR.Type]
+liftNonRecFuncArgTypes = map liftType
+
+-- | Converts the argument types of a recursive function using @convertType@.
+--
+--   The outermost argument at the given index is marked decreasing.
+liftRecFuncArgTypes :: Int -> [IR.Type] -> [LIR.Type]
+liftRecFuncArgTypes decIndex args =
+  let convArgs                      = map liftType args
+      (startArgs, decArg : endArgs) = splitAt (decIndex - 1) convArgs
+  in  startArgs ++ (markOutermostDecreasing decArg : endArgs)
+
+-- | Converts a constructor argument using @convertType@.
+--
+--   All occurrences of the constructed type are marked as structurally smaller.
+liftConArgType :: IR.QName -> IR.Type -> LIR.Type
+liftConArgType ident = markAllDec ident . liftType
+
+-------------------------------------------------------------------------------
+-- Translations                                                              --
+-------------------------------------------------------------------------------
+
+-- | Converts a type from IR to lifted IR by lifting it into the @Free@ monad.
+--
+--   This corresponds to the dagger translation for monotypes as described by
+--   Abel et al.
+--
+--   > τ' = Free τ*
+liftType :: IR.Type -> LIR.Type
+liftType = LIR.FreeTypeCon NoSrcSpan . liftType'
+
+-- | Lifts a type from IR to lifted IR by adding the free arguments to
+--   constructors and lifting function types in the @Free@ monad.
+--
+--   This corresponds to the star translation for monotypes as described by
+--   Abel et al.
+--
+--   > (τ₁ τ₂)* = τ₁* τ₂*
+--   > (τ₁ -> τ₂)* = τ₁' -> τ₂'
+--   > C* = Ĉ Shape Position
+--   > α* = α̂
+liftType' :: IR.Type -> LIR.Type
+liftType' (IR.TypeCon srcSpan name) = LIR.TypeCon srcSpan name Nil False
+liftType' (IR.TypeVar srcSpan name) = LIR.TypeVar srcSpan name
+liftType' (IR.TypeApp _ l r       ) = liftType' l `LIR.typeApp` liftType' r
+liftType' (IR.FuncType srcSpan l r) =
+  LIR.FuncType srcSpan (liftType l) (liftType r)
```

Comment:
My problem with this approach is approach is that it also doesn't solve the problem and just moves it. Storing the type parameters in the 'correct' order forces us to use `foldl` to build the AST in the backend, therefore just reversing the list again. Even worse, the IR already has the right structure! This approach just reverses the structure twice while passing it through the pipeline.
```
     ⋰                                        ⋰
    @                         :              $  
   / \                       / \            / \ 
  @   τ₂   =>  IR.TypeApp C τ₁  :    =>    $   τ₂  
 / \                           / \        / \   
C   τ₁                        τ₂  …      C   τ₁ 
```
Storing the list in 'reverse' order (or as a snoc list to just visually flip it) doesn't change the structure, allowing us to use the right recursion scheme to build the AST in the back end (`foldr` for `[LIR.Type]` or `foldl` for `SnocList LIR.Type`), building the AST without reversing the list. 

Since parsing and pretty printing are left to right operations we have to deal with the problem of reversing the list at some point. I think we should just solve the problem as early as possible to avoid influencing the main flow of data (i.e. reversing all type applications twice while compiling).

Dealing with left recursion while parsing shouldn't be a problem. We can either parse the structure right recursive and reverse it later (`reverse` or `foldl`) or use [`chainl`](https://hackage.haskell.org/package/parsec-3.1.14.0/docs/src/Text.Parsec.Combinator.html#chainl), which avoids unnecessary backtracking by parsing bottom up. Note that dealing with the problem while parsing avoids the `foldl` in the backend.

While pretty printing we of course have to deal with the problem again. One way of reducing the thunk buildup of `foldl` is to use [`foldl'`](https://hackage.haskell.org/package/base-4.14.0.0/docs/Data-List.html#v:foldl-39-) instead.

---

From a performance point of view (which shouldn't really matter, because our list has like 5 items :smile:) it's just a trade of between reversing during IO or between pipeline stages. 

From an elegance/readability point of view I'm not quite sure which approach is better. I think I lean towards the simpler algorithms working on the expected structure. As a side note: I already started working on the conversion of expressions. The Coq backend uses the same technique of accumulating arguments which I reused, because in this case it simplifies the algorithm.

Reply:
> performance [...] shouldn't really matter, because our list has like 5 items

I agree. Performance is not really a priority for us. But why do we then care about the implicit reversals performed by `foldl` in the back end? The point I wanted to make with my performance argument is that the `SnocList` approach complicates the code but doesn't even improve performance.

> Since parsing and pretty printing are left to right operations we have to deal with the problem of reversing the list at some point.

I've included the parsing and pretty printing example to show that even though `liftType'` can be defined slightly more compactly with `SnocList`s, other code will become more complex. For parsing a list of arguments, we can simply use `Parsec.many` and `hsep . map pretty` for pretty printing. I guess that much more code is needed to parse or pretty print a `SnocList`.

> The Coq backend uses the same technique of accumulating arguments

That's only partially true. Yes, I don't use an accumulator argument in [`convertType'`](https://github.com/FreeProving/free-compiler/blob/35bbb90b0c5d776489dae00539b2b55acbfb0ae1/src/lib/FreeC/Backend/Coq/Converter/Type.hs#L44). However, unlike `Agda.typeApp`, [`Coq.app`](https://github.com/FreeProving/free-compiler/blob/35bbb90b0c5d776489dae00539b2b55acbfb0ae1/src/lib/FreeC/Backend/Coq/Syntax.hs#L92) adds the type argument to the back of the list and not to the front. If I recommended to use `(++)` instead of `(:)` in `Agda.typeApp`, I'd most likely have started a discussion about performance. That's one reason why I've suggested to use an accumulator argument. The other reason is that [`convertExpr'`](https://github.com/FreeProving/free-compiler/blob/35bbb90b0c5d776489dae00539b2b55acbfb0ae1/src/lib/FreeC/Backend/Coq/Converter/Expr.hs#L41) requires an accumulator argument. Therefore, adding an accumulator argument to `liftType'` improves consistency.